docs: fix links that pointed to earlier Juju docs

dwilding opened the pull request:
This PR replaces links to https://juju.is/docs/ by internal cross-references to other Ops pages or intersphinx cross-references to Juju pages (as appropriate).

**Not included in this PR**

- The link in https://ops.readthedocs.io/en/latest/howto/get-started-with-charm-testing.html#charmcraft-profiles that should point to the Charmcraft docs. I'll follow up with a separate PR after https://github.com/canonical/charmcraft/pull/2085 gets merged into Charmcraft.

- Links in the code samples of the Kubernetes tutorial (for example, in [Create a minimal Kubernetes charm](https://ops.readthedocs.io/en/latest/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/create-a-minimal-kubernetes-charm.html)). I think it would be better to tackle these separately when improving the tutorial - @tonyandrewmeyer what do you think?

dwilding commented on docs/howto/get-started-with-charm-testing.md:
```diff
@@ -1,4 +1,3 @@
-(get-started-with-charm-testing)=
```
Having a label that's the same as the doc title causes ambiguity when creating a cross reference.

dwilding commented on docs/explanation/how-and-when-to-defer-events.md:
```diff
@@ -1,4 +1,3 @@
-(how-and-when-to-defer-events)=
```
Having a label that's the same as the doc title causes ambiguity when creating a cross reference.

dwilding commented on docs/howto/turn-a-hooks-based-charm-into-an-ops-charm.md:
```diff
@@ -136,7 +136,6 @@ It is in our interest to move the handler logic for each `/hooks/<hook_name>` to
 - We can avoid code duplication by accessing shared data via the CharmBase interface provided through `self`. 
 - The code is all in one place, easier to maintain.
 - We automatically have one Python object we can test, instead of going back and forth between Bash scripts and Python wrappers.
-- We can use [the awesome testing Harness](https://juju.is/docs/sdk/testing).
```
Remove this bullet because the preceding bullet already mentions testing. And Harness is not considered so awesome any more :broken_heart:

dwilding replied:
I think it's best to omit the link here, as it's not something we want the reader to start working through right away. I'm adding a small section towards the end of the doc to remind the reader that they should write tests, with a link to [Get started with charm testing](https://ops.readthedocs.io/en/latest/howto/get-started-with-charm-testing.html).

dwilding commented on docs/explanation/how-and-when-to-defer-events.md:
```diff
@@ -1,4 +1,3 @@
-(how-and-when-to-defer-events)=
 # How, and when, to defer events
```
Having a label that's the same as the doc title causes ambiguity when creating a cross reference.

dwilding commented on docs/howto/get-started-with-charm-testing.md:
```diff
@@ -1,4 +1,3 @@
-(get-started-with-charm-testing)=
 # Get started with charm testing
```
Having a label that's the same as the doc title causes ambiguity when creating a cross reference.

dwilding commented on docs/howto/run-workloads-with-a-charm-kubernetes.md:
```diff
@@ -204,7 +204,7 @@ See the [layer specification](https://canonical-pebble.readthedocs-hosted.com/en
 #### Add a configuration layer
 
-To add a configuration layer, call [`Container.add_layer`](ops.Container.add_layer) with a label for the layer, and the layer's contents as a YAML string, Python dict, or [`pebble.Layer`](#ops.pebble.Layer) object.
+To add a configuration layer, call [`Container.add_layer`](ops.Container.add_layer) with a label for the layer, and the layer's contents as a YAML string, Python dict, or [`pebble.Layer`](ops.pebble.Layer) object.
```
This change is out of scope of this PR, but it's the only instance that needs adjusting for consistency, so I'm including it here.

dwilding commented on docs/howto/manage-logs.md:
```diff
@@ -35,7 +35,7 @@ class HelloOperatorCharm(ops.CharmBase):
 > - [`logging.getLogger().debug()`](https://docs.python.org/3/library/logging.html#logging.Logger.debug)
 
-Juju automatically picks up logs from charm code that uses the Python [logging facility](https://docs.python.org/3/library/logging.html), so we can use the Juju [`debug-log` command](https://juju.is/docs/juju/juju-debug-log) to display logs for a model. Note that it shows logs from the charm code (charm container), but not the workload container. Read ["Use `juju debug-log`"](https://juju.is/docs ...
+Juju automatically picks up logs from charm code that uses the Python [logging facility](https://docs.python.org/3/library/logging.html), so we can use the Juju [`debug-log` command](inv:juju:std:label#command-juju-debug-log) to display logs for a model. Note that it shows logs from the charm code (charm container), but not the workload container.
```
I removed the link to https://juju.is/docs/sdk/get-logs-from-a-kubernetes-charm#heading--use-juju-debug-log because I can't find where this info has ended up. @tmihoc do you know whether we still have it somewhere?

dwilding replied:
I don't think [Juju | How to manage logs](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-logs/) addresses the point in such an explicit way. Anyway, I agree it's a more appropriate place to link to. Good idea about "see also" - I'm putting it in a "see more", to be consistent with the rest of the page.

tmihoc replied:
I'd use a `>See first: Juju | Manage logs'. I've done that for both Charmcraft and Ops. The idea being that we want people to be aware of what a thing looks like on the charm user end _first_, as that's likely to be the fastest way for them to catch up.

dwilding replied:
Great point. I just realised that we already have a "see first" link to that page. I've updated the "see first" block to list the how-to link before the reference link.

dwilding replied:
Oops, I just found the commit for this sitting in my editor. I thought I'd pushed it, but apparently not... I'll do a separate PR for it.

dwilding commented on docs/howto/get-started-with-charm-testing.md:
```diff
@@ -167,7 +167,7 @@ def test_pebble_ready_writes_config_file():
 
 Where unit testing focuses on black-box method-by-method verification, integration testing focuses on the big picture. Typically integration tests check that the charm does not break (generally this means: blocks with status `blocked` or `error`) when a (mocked) cloud admin performs certain operations. These operations are scripted by using, in order of abstraction:
- -  shell commands against [the `juju` cli](https://juju.is/docs/olm/juju-cli-commands)
+ -  shell commands against [the `juju` CLI](inv:juju:std:label#list-of-juju-cli-commands)
```
I'm using the MyST format for the intersphinx link to allow backticks in the link title.

dwilding commented on docs/howto/manage-relations.md:
```diff
@@ -123,11 +123,11 @@ To use data received through the relation, have your charm observe the `relation
 Most of the time, you should use the same holistic handler as when receiving other data, such as `secret-changed` and `config-changed`. To access the relation(s) in your holistic handler, use the [](ops.Model.get_relation) method or [](ops.Model.relations) attribute.
 
-> See also: {ref}`holistic-vs-delta-charms`
+> See also: [](/explanation/holistic-vs-delta-charms)
```
This change is out of scope of this PR, but I'm updating another instance, so using the same format here.

dwilding commented on docs/howto/run-workloads-with-a-charm-machines.md:
```diff
@@ -2,9 +2,6 @@
 # How to run workloads with a charm - machines
 
 There are several ways your charm might start a workload, depending on the type of charm you’re authoring. 
-<!--
-Before writing the code to start your workload, recall the [Lifecycle events](https://juju.is/docs/sdk/events) section, and note that when the `start` event is emitted, charm authors should ensure their workloads are configured to "persist in a started state without further intervention from Juju or an administrator".
--->
```
I'm not sure why this is commented out, but might as well remove it.

tonyandrewmeyer commented on docs/howto/turn-a-hooks-based-charm-into-an-ops-charm.md:
````diff
@@ -206,7 +205,7 @@ That allows us to fetch the Relation wherever we need it and access its contents
 ```
 
-Note how `relation.data` provides an interface to the relation databag (more on that [here](https://juju.is/docs/sdk/relations#heading--relation-data)) and we need to select which part of that bag to access by passing an `ops.model.Unit` instance.
+Note how `relation.data` provides an interface to the relation databag (see [](ops.Relation.data)) and we need to select which part of that bag to access by passing an `ops.model.Unit` instance.
````
This was going to [a Juju reference page previously](https://web.archive.org/web/20221102151915/https://juju.is/docs/sdk/integration#heading--integration-data) rather than to the ops API reference. I can't figure out where that is now: @tmihoc did it get removed?

If it doesn't exist any more, I think maybe linking to the section in the relation how-to might be better here?

tonyandrewmeyer commented on docs/tutorial/write-your-first-machine-charm.md:
```diff
@@ -522,7 +522,7 @@ Congratulations, your charm user can view the version of the workload deployed f
 ## Tear things down
 
-> See [Juju | Tear down your development environment automatically](https://juju.is/docs/juju/set-up--tear-down-your-test-environment#tear-down-automatically)
+> See the automatic teardown instructions in {external+juju:ref}`Juju | Manage your deployment environment <manage-your-deployment-environment>`.
```
Is it possible to link directly to the tear down section? If not that's fine.

dwilding replied:
Yeah, I agree that would be better. I've asked Dora about adding a target so that we can directly link there.

tmihoc replied:
@dwilding Added it in this PR: https://github.com/juju/juju/pull/18911/files

dwilding replied:
Thank you @tmihoc. Do you know when the targets will be available in the `latest` version of the Juju docs?

dwilding replied:
All good now

james-garner-canonical commented on docs/explanation/how-and-when-to-defer-events.md:
```diff
@@ -33,7 +33,7 @@ In some situations, the charm is waiting for a system to be ready, but it’s no
 Deferring the work here is ok, but it’s important to consider the delay between deferring the event and its eventual re-emitting - it’s not safe to assume that this will be a small period of time, unless you know that another event can be expected.
 
-For a Kubernetes charm, If the charm is waiting on the workload and it’s possible to have the workload execute a command when it’s ready, then using a [Pebble custom notice](https://juju.is/docs/sdk/interact-with-pebble#heading--use-custom-notices-from-the-workload-container) is much better than deferring. This then becomes another example of “waiting for a collection of events”, described above.
+For a Kubernetes charm, If the charm is waiting on the workload and it’s possible to have the workload execute a command when it’s ready, then using a [Pebble custom notice](#use-custom-notices-from-the-workload-container) is much better than deferring. This then becomes another example of “waiting for a collection of events”, described above.
```
Capital 'If' after first comma

dwilding replied:
Fixed in 65bbef2, thanks!

james-garner-canonical reviewed (approved):
Look good! I built the docs locally and went through to check that the links all worked. Approving as I don't really have anything significant to add over Tony's review.

tonyandrewmeyer commented:
> I built the docs locally and went through to check that the links all worked.

I did the same with the preview build on rtd, by the way. So between us we have good coverage :)

james-garner-canonical commented:
> > I built the docs locally and went through to check that the links all worked.
> 
> I did the same with the preview build on rtd, by the way. So between us we have good coverage :)

There's a preview build!?

dwilding commented:
> There's a preview build!?

The preview build is here: https://ops--1575.org.readthedocs.build/en/1575/. You can find it by clicking **Details** for the check called "docs/readthedocs.org:ops". Sorry, I should have pasted it into the PR description (I usually do that).

dwilding commented:
Thanks everyone for your feedback! @tonyandrewmeyer I've addressed all the comments now - please could you review again when you get a chance?

One thing to note: while updating the links to [Set things up](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-your-deployment/manage-your-deployment-environment/#set-things-up) and [Tear things down](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-your-deployment/manage-your-deployment-environment/#tear-things-down), I also made some formatting changes to smooth out the flow (to my eyes). These changes: https://github.com/canonical/operator/pull/1575/commits/5f4828ac1be428818be2df63e1d0dd58b6736000

tonyandrewmeyer reviewed (approved):
Thanks - looking great!
